lib/service/chat_service.dart

```diff
@@ -0,0 +1,223 @@
+        _id == id &&
+        _wallet == wallet &&
+        _websocketChannel != null) {
+      log.info("[CHAT] connect to the same channel. Do nothing");
+      return;
+    }
+    _address = address;
+    _id = id;
+    _wallet = wallet;
+    final link = "/v1/chat/ws?index_id=$id&address=$address";
+    final header = await _getHeader(link, wallet, address);
+    _websocketChannel = IOWebSocketChannel.connect(
+      "${Environment.postcardChatServerUrl}$link",
+      headers: header,
+      customClient: HttpClient(),
+      pingInterval: const Duration(seconds: 50),
+    );
+    await _websocketChannel?.ready;
+
+    // listen events
+
+    _websocketChannel?.stream.listen(
+      (event) {
+        log.info("[CHAT] event: $event");
+        final response = app.WebsocketMessage.fromJson(json.decode(event));
+        switch (response.command) {
+          case 'NEW_MESSAGE':
+            try {
+              final newMessages = (response.payload["messages"]
+                      as List<dynamic>)
+                  .map((e) => app.Message.fromJson(e as Map<String, dynamic>))
+                  .toList();
+              for (var element in _listeners) {
+                element.onNewMessages(newMessages);
+              }
+            } catch (e) {
+              log.info("[CHAT] NEW_MESSAGE error: $e");
+            }
+            break;
+          case 'RESP':
+            if (response.payload["ok"] != null &&
+                response.payload["ok"].toString() == "1") {
+              for (var element in _listeners) {
+                element.onResponseMessage(response.id, ChatService.SENT);
+              }
+            } else if (response.payload["error"] != null) {
+              for (var element in _listeners) {
+                element.onResponseMessage(response.id, ChatService.ERROR);
+              }
+            } else {
+              try {
+                final newMessages =
+                    (response.payload["messages"] as List<dynamic>)
+                        .map((e) => app.Message.fromJson(e))
+                        .toList();
+                for (var element in _listeners) {
+                  element.onResponseMessageReturnPayload(
+                      newMessages, response.id);
+                }
+              } catch (e) {
+                log.info("[CHAT page] RESP error: $e");
+              }
+            }
+            break;
+          default:
+            break;
+        }
+      },
+      onDone: () async {
+        log.info("[CHAT] onDone");
+
+        Future.delayed(const Duration(seconds: 5), () async {
+          if (_address != null && _id != null && _wallet != null) {
+            _address = null;
+            _id = null;
+            _wallet = null;
+            log.info("[CHAT] _websocketChannel reconnecting");
+            await connect(
+              address: _address!,
+              id: _id!,
+              wallet: _wallet!,
+            );
+            for (var element in _listeners) {
```

Comment:
why need to re-add the listeners? the WS is closed but the listeners are still there?

Reply:
in case of dispose, it won't go in there.
this is the case for reconnect if internet connection disrupts